.github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py

```diff
@@ -1,10 +1,135 @@
 from collections import defaultdict
+import functools
+import math
 import numpy as np
 import pandas as pd
 
+from .protobuf import xla_module_metadata
 from .utils import make_child_mask
 
 
+def element_type_in_bits(element_type: int) -> int:
+    """
+    Given an int representing an XLA PrimitiveType enum value, return the width of that
+    type in bits.
+    """
+    from xla.xla_data_pb2 import PrimitiveType
+
+    enum_name = PrimitiveType.Name(element_type)
+    if enum_name == "PRED":
+        # Based on
+        # https://github.com/openxla/xla/blob/664a36a2b5e5be9179c5841830da56799b6dfe60/xla/service/gpu/runtime/nccl_api.cc#L116-L118
+        return 8
+
+    # S32 is a 32-bit type and so on.
+    # FIXME: does not handle FP8 yet
+    for prefix in ["BF", "C", "F", "S", "U"]:
+        if enum_name.startswith(prefix):
+            return int(enum_name[len(prefix) :])
+    raise Exception(f"Could not deduce size of {enum_name}")
+
+
+@functools.lru_cache
+def get_message_size(program_id: int, instruction: str) -> int:
```

Comment:
Returns a dataframe, not an int, no?

Reply:
Right 🤦  -- all the more reason to actually set up some CI to lint this.